Source/Shared/Storage/ReadSyncWriteAsyncStorage.swift

```diff
@@ -0,0 +1,103 @@
+import Foundation
+
+/// Manipulate storage in a "read sync, write async" manner.
+final class ReadSyncWriteAsyncStorage {
+  let internalStorage: StorageAware
+  let concurrentQueue = DispatchQueue(label: "Cache.ReadSyncWriteAsyncStorage.Queue",
+                                      attributes: .concurrent)
+
+  init(storage: StorageAware) {
+    self.internalStorage = storage
+  }
+}
+
+extension ReadSyncWriteAsyncStorage {
+  public func entry<T: Codable>(forKey key: String) throws -> Entry<T> {
+    var entry: Entry<T>!
+    try concurrentQueue.sync {
+      entry = try internalStorage.entry(forKey: key) as Entry<T>
+    }
+
+    return entry
+  }
+
+  func object<T: Codable>(forKey key: String) throws -> T {
+    return try entry(forKey: key).object
+  }
+
+  func existsObject<T: Codable>(ofType type: T.Type, forKey key: String) throws -> Bool {
+    do {
```

Comment:
What about `try? object(forKey: key) != nil`

Reply:
@vadymmarkov you need to cast to the correct type